### DISCUSSION TITLE
MEEP: Consider `@GraphPrivate` annotation to prevent leakage of certain types

### DISCUSSION THREAD

**Discussion by ZacSweers** [ZacSweers/metro#1769] [MEEPs]
This a proposal for a new `@ConfineToGraph` annotation, which would confine the annotated binding _to the graph it's bound in_. This is useful for situations where you have an extendable `@DependencyGraph` or `@GraphExtension` but want to prevent certain bindings from being implicitly shared to child graphs.

## Example

Consider a scenario where you have some base `HttpClient` on your `AppGraph` that you use in multiple places in your app graph but do not want to expose to extensions.

Using `@ConfineToGraph`, you can confine it just to the `AppGraph`.

```kotlin
@DependencyGraph(AppScope::class)
interface AppGraph {
  @Provides
  @ConfineToGraph
  @SingleIn(AppScope::class)
  fun provideUnauthedHttpClient(): HttpClient = HttpClient(...)

  val loggedInGraph: LoggedInGraph
}

@GraphExtension
interface LoggedInGraph {
  // Error!
  val httpClient: HttpClient
}
```

Not applicable to the following scenarios, as these are always looked up lazily:
- Unscoped constructor-injected class bindings
    - If you do want to limit a class, provide it explicitly with Provides instead of constructor-injecting it.
- `@AssistedFactory`
- `MembersInjector`

This only prevents its direct visibility to child graphs. It does not (and in fact is to encourage) indirect usage, as this is purely for enforcing encapsulation of that binding.

If a graph extension does try to use this binding, it will get a `MissingBinding` error message with a hint that informs of its confined nature.

```
[Metro/MissingBinding] Could not find a binding for HttpClient!
  ...

(Hint)
HttpClient exists in parent graph AppGraph but is confined to that graph.
```

## Open questions

1. Is this valuable?
2. Is the name right? Or should it be `@EncapsulateWithin` or something more explicit?
3. Should it be allowed on classes at all? Or should this only be limited to `@Provides`/`@*Binds*` declarations?
4. Should we expose a way to undo this, perhaps via exposing an accessor?
  - As opposed to just removing the confinement annotation entirely

**Comment by jingibus** [comment]
1. Yes, I think so. Another scenario it would be really useful for would be providing scope-specific bindings — e.g. we could have dependencies inject `coroutineScope: CoroutineScope`. And then in `AppScope` we bind it to a global coroutine scope, in `SignedInScope` a coroutine scope that is only live while signed in, etc
2. `@GraphPrivate`?
3. It's 1:1 with a binding, right? So yeah, I think it gets confusing if it's ever sited separately from an annotation that produces a binding. It would be worth considering introducing it as an annotation parameter instead of a separate annotation.
4. I could see us using existing APIs to expose private deps with qualifiers more broadly. E.g.:

```kotlin
@Binds @ConfineToGraph fun myDep(myDep: RealMyDep): MyDep

// expose with an additional qualifier for subgraphs
@Binds @ForScope(AppScope::class) fun myDepGlobal(myDep: MyDep): MyDep
```

For us, I don't see an API for "undoing" it as usefult, though. 

Another question: does this apply to multibindings? And if so, how? (I'm guessing that it would need to be applicable to both the multibinding set itself and to the individual bindings)

**Comment by Egorand** [comment]
Would it be possible/desirable to confine deps to scopes rather than graphs? In that case, there could potentially be a `@PrivateIn` annotation with same annotation targets as `@SingleIn`, but which would make the dependency inaccessible outside of that scope.

**Reply by ZacSweers** [reply]
I thought about that but I worry it conflates this too much with actual scoping since you can use this on unscoped bindings (or with this we would only use it on scoped types).

**Comment by mecharip** [comment]
I'd be a fan of this, as currently I have `Scoped` interface as means to get lifecycle events for the (any) given scope its plugged into, and now I need to disambiguate with qualifiers. It works but is not as nice

**Comment by ZacSweers** [comment]
This is now implemented in #1877 and will be in Metro 0.11.0 as an experimental feature

**Comment by ZacSweers** [comment]
A followup I have to this - should extensions now no longer implicitly override bindings of the same key in the parent? i.e. should we impose using GraphPrivate to break the duplicate binding potential?

**Comment by ZacSweers** [comment]
This is stable in 1.2.0
